src/Compilers/CSharp/Portable/BoundTree/VariablePendingInference.cs

```diff
@@ -63,26 +79,39 @@ private BoundExpression SetInferredType(TypeSymbol type, Binder binderOpt, Diagn
 
                     return new BoundFieldAccess(this.Syntax,
                                                 this.ReceiverOpt,
-                                                fieldSymbol, null, LookupResultKind.Viable, type, 
+                                                fieldSymbol, null, LookupResultKind.Viable, type,
                                                 this.HasErrors || inferenceFailed);
 
                 default:
                     throw ExceptionUtilities.Unreachable;
             }
+        }
 
+        internal BoundExpression FailInference(Binder binder, DiagnosticBag diagnosticsOpt)
+        {
+            return this.SetInferredType(null, binder, diagnosticsOpt);
         }
 
         private void ReportInferenceFailure(DiagnosticBag diagnostics)
         {
-            var designation = (SingleVariableDesignationSyntax)((DeclarationExpressionSyntax)this.Syntax).Designation;
+            SingleVariableDesignationSyntax designation;
+            switch (this.Syntax.Kind())
+            {
+                case SyntaxKind.DeclarationExpression:
+                    designation = (SingleVariableDesignationSyntax)((DeclarationExpressionSyntax)this.Syntax).Designation;
+                    break;
+                case SyntaxKind.SingleVariableDesignation:
+                    designation = (SingleVariableDesignationSyntax)this.Syntax;
+                    break;
+                default:
+                    throw ExceptionUtilities.Unreachable;
+            }
+
             Binder.Error(
-                diagnostics, ErrorCode.ERR_TypeInferenceFailedForImplicitlyTypedOutVariable, designation.Identifier,
+                diagnostics, this.InferenceFailedError, designation.Identifier,
                 designation.Identifier.ValueText);
         }
```

Comment:
Perhaps make `ReportInferenceFailure` an abstract method, and remove `InferenceFailedError`. (It seems like a coincidence that the two errors have the same arguments and location rather than an opportunity to share code here.)